### PR TITLE
Add gradients to the scatter_max and scatter_min operations.

### PR DESCRIPTION
**Pull request by alexdavies**
This is being done to allow the creation of a differentiable segment_max. Segment_max is an important operation for GraphNets and is an open feature request at https://github.com/google/jax/issues/2255

### CODE REVIEW

**Comment by alexdavies** [comment]
@googlebot I signed it!

File: jax/lax/lax.py [hunk 1]
```diff
@@ -3685,20 +3685,90 @@ def _scatter_mul_jvp_rhs(g, x, i, y, *, dimension_numbers, **kw):
...
+
+    coef = select((actual_updates == 0) | (summed_updates == 0),
+                  _ones(updates),
+                  actual_updates / summed_updates)
+
+    # This can be simplified once scatter has transpose implemented
+    target_tangents = gather(
+        g_operand, scatter_indices, gather_dnums, onp.array(slice_sizes))
+
+    tangent_updates = select(updates == target_vals,
```
Comment (alexdavies) [inline jax/lax/lax.py:3748]: TODO: This actually won't correctly account for the case where an updates value is scattered to a location with the same value.

Comment (jekbradbury) [reply jax/lax/lax.py:3748]: Does this TODO mean the PR isn't ready for review, or would you like to get it in without this support first?

Comment (alexdavies) [reply jax/lax/lax.py:3748]: Yes - hold the review, I'll fix this on Monday as it only occurred to me after I'd submitted. It won't fundamentally change the CL though so any general comments are welcome.

**Comment by alexdavies** [comment]
I updated the PR to fix the previous issue, it's ready for review now.

File: jax/lax/lax.py [hunk 2]
```diff
@@ -3685,20 +3685,113 @@ def _scatter_mul_jvp_rhs(g, x, i, y, *, dimension_numbers, **kw):
...
+      if i in scatter_dnums.inserted_window_dims:
+        slice_sizes.append(1)
+      else:
+        slice_sizes.append(updates_shape[scatter_dnums.update_window_dims[pos]])
+        pos += 1
+
+    # For consistency with other max operations, if there are two or more values
+    # in updates that are contending to replace the same index location, the
+    # resulting tangent at that location will be the average of the associated
+    # tangents for the values in updates.
```
Comment (jekbradbury) [inline jax/lax/lax.py:3725]: With #3147 we'll be able to avoid this step if we know indices aren't duplicated.

**Review by jekbradbury — APPROVED** [review]

**Comment by jekbradbury** [comment]
This looks great. I'm going to merge now, but we'll probably want to update this rule a little when #3147 goes in (that PR is mainly there to trigger optimized XLA:TPU code in certain large scatter cases).

**Pull request merged**